File: src/calculus_u_substitution_easy.py
"""
calculus - u_substitution (easy)
Generated: 2026-02-22T05:12:24.203042
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # ∫ (ax + b)^n dx   a in {±1..±5}, b in {-8..8}, n in {2..6}
        a_val = nonzero(-5, 5)
        b_val = randint(-8, 8)
        n_val = randint(2, 6)

        u_expr = a_val * x + b_val
        integrand = u_expr**n_val
        ans = (u_expr**(n_val + 1)) / (a_val * (n_val + 1))

        question = f"\\int {latex(integrand)} \\, dx"

        sol = steps(
            f"Let $u = {latex(u_expr)}$",
            f"Then $du = {a_val} \\, dx$, so $dx = \\frac{{1}}{{{a_val}}} \\, du$",
            f"The integral becomes $\\frac{{1}}{{{a_val}}} \\int u^{{{n_val}}} \\, du$",
            f"$= \\frac{{1}}{{{a_val}}} \\cdot \\frac{{u^{{{n_val+1}}}}}{{{n_val+1}}} + C$",
            f"$= {latex(ans)} + C$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1000, 1150),
            topic="calculus/u_substitution",
            solution=sol
        )

    elif problem_type == 2:
        # ∫ x^m * (a*x^(m+1) + c)^n dx   — u = inner poly
        # Keep m=1 for easy level, n in {2..5}, a in {1..4}, c in {-6..6}
        a_val = randint(1, 5)
        c_val = randint(-6, 6)
        n_val = randint(2, 5)

        u_expr = a_val * x**2 + c_val
        integrand = x * (u_expr)**n_val
        # u = a*x² + c, du = 2a*x dx  →  x dx = du/(2a)
        ans = (u_expr**(n_val + 1)) / (2 * a_val * (n_val + 1))

        question = f"\\int {latex(integrand)} \\, dx"

        sol = steps(
            f"Let $u = {latex(u_expr)}$",
            f"Then $du = {2*a_val}x \\, dx$, so $x \\, dx = \\frac{{1}}{{{2*a_val}}} \\, du$",
            f"The integral becomes $\\frac{{1}}{{{2*a_val}}} \\int u^{{{n_val}}} \\, du$",
            f"$= \\frac{{u^{{{n_val+1}}}}}{{{2*a_val*(n_val+1)}}} + C$",
            f"$= {latex(ans)} + C$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1100, 1250),
            topic="calculus/u_substitution",
            solution=sol
        )

    elif problem_type == 3:
        # ∫ sin(ax + b) dx  or  ∫ cos(ax + b) dx
        a_val = nonzero(-5, 5)
        b_val = randint(-6, 6)
        trig_choice = choice(['sin', 'cos'])
        u_expr = a_val * x + b_val

        if trig_choice == 'sin':
            integrand = sin(u_expr)
            ans = -cos(u_expr) / a_val
            antideriv_u = "-\\cos(u)"
        else:
            integrand = cos(u_expr)
            ans = sin(u_expr) / a_val
            antideriv_u = "\\sin(u)"

        question = f"\\int {latex(integrand)} \\, dx"

        sol = steps(
            f"Let $u = {latex(u_expr)}$",
            f"Then $du = {a_val} \\, dx$, so $dx = \\frac{{1}}{{{a_val}}} \\, du$",
            f"The integral becomes $\\frac{{1}}{{{a_val}}} \\int {trig_choice}(u) \\, du = \\frac{{1}}{{{a_val}}} \\cdot {antideriv_u} + C$",
            f"$= {latex(ans)} + C$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1000, 1200),
            topic="calculus/u_substitution",
            solution=sol
        )

    elif problem_type == 4:
        # ∫ e^(ax + b) dx   a in {±1..±5}, b in {-5..5}
        a_val = nonzero(-5, 5)
        b_val = randint(-5, 5)

        u_expr = a_val * x + b_val
        integrand = exp(u_expr)
        ans = exp(u_expr) / a_val

        question = f"\\int {latex(integrand)} \\, dx"

        sol = steps(
            f"Let $u = {latex(u_expr)}$",
            f"Then $du = {a_val} \\, dx$, so $dx = \\frac{{1}}{{{a_val}}} \\, du$",
            f"$\\frac{{1}}{{{a_val}}} \\int e^u \\, du = \\frac{{1}}{{{a_val}}} e^u + C$",
            f"$= {latex(ans)} + C$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1050, 1250),
            topic="calculus/u_substitution",
            solution=sol
        )

    else:
        # ∫ x^2 * (a*x^3 + b)^n dx   n in {2..5}, a in {1..4}, b in {-5..5}
        a_val = randint(1, 5)
        b_val = randint(-5, 5)
        n_val = randint(2, 5)

        u_expr = a_val * x**3 + b_val
        integrand = x**2 * u_expr**n_val
        # u = a*x³ + b, du = 3a*x² dx  →  x² dx = du/(3a)
        ans = (u_expr**(n_val + 1)) / (3 * a_val * (n_val + 1))

        question = f"\\int {latex(integrand)} \\, dx"

        sol = steps(
            f"Let $u = {latex(u_expr)}$",
            f"Then $du = {3*a_val}x^2 \\, dx$, so $x^2 \\, dx = \\frac{{1}}{{{3*a_val}}} \\, du$",
            f"$\\frac{{1}}{{{3*a_val}}} \\int u^{{{n_val}}} \\, du = \\frac{{u^{{{n_val+1}}}}}{{{3*a_val*(n_val+1)}}} + C$",
            f"$= {latex(ans)} + C$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1100, 1250),
            topic="calculus/u_substitution",
            solution=sol
        )

emit(generate())

File: src/calculus_u_substitution_easy.jl
# calculus - u_substitution (easy)
# Generated: 2026-03-08T20:45:22.531524

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x u begin
    set_topic!("calculus/u_substitution")
    
    # Choose problem type randomly
    problem_type = choice(1:4)
    
    if problem_type == 1
        # Type 1: ∫ (ax+b)^n dx where n ≠ -1
        # Pick clean answer first
        a = nonzero(-5, 5)
        b = randint(-15, 15)
        n = choice([2, 3, 4, 5])
        
        inner = a*x + b
        integrand = inner^n
        
        # Antiderivative: (ax+b)^(n+1) / (a(n+1)) + C
        antideriv = inner^(n+1) / (a*(n+1))
        
        problem(
            question="Evaluate \\int $(tex(integrand)) \\, dx",
            answer=antideriv,
            difficulty=(700, 900),
            solution=steps(
                sol("Given", integrand),
                "Let u = $(tex(inner)), then du = $(a) dx",
                "Rewrite: \\int u^{$(n)} \\cdot \\frac{1}{$(a)} du = \\frac{1}{$(a)} \\cdot \\frac{u^{$(n+1)}}{$(n+1)} + C",
                sol("Substitute back", antideriv)
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Type 2: ∫ f'(x) * [f(x)]^n dx
        # Example: ∫ 2x(x²+c)^n dx
        a = choice([2, 3, 4])
        c = randint(-10, 10)
        n = choice([2, 3, 4])
        
        inner = x^2 + c
        integrand = a*x * inner^n
        
        # u = x² + c, du = 2x dx
        # Answer: (x²+c)^(n+1) / (2(n+1)) * a + C
        antideriv = inner^(n+1) * a / (2*(n+1))
        
        problem(
            question="Evaluate \\int $(tex(integrand)) \\, dx",
            answer=antideriv,
            difficulty=(800, 1000),
            solution=steps(
                sol("Given", integrand),
                "Let u = $(tex(inner)), then du = $(tex(2*x)) dx",
                "Rewrite: \\int u^{$(n)} \\cdot \\frac{$(a)}{2} du = \\frac{$(a)}{2} \\cdot \\frac{u^{$(n+1)}}{$(n+1)} + C",
                sol("Substitute back", antideriv)
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Type 3: ∫ 1/(ax+b) dx
        a = nonzero(-6, 6)
        b = randint(-12, 12)
        
        inner = a*x + b
        integrand = 1/inner
        
        # Answer: ln|ax+b|/a + C
        # We'll format as string since ln needs special handling
        answer_str = "\\frac{1}{$(a)}\\ln|$(tex(inner))|"
        
        problem(
            question="Evaluate \\int $(tex(integrand)) \\, dx",
            answer=answer_str,
            difficulty=(900, 1100),
            answer_type="expression",
            solution=steps(
                sol("Given", integrand),
                "Let u = $(tex(inner)), then du = $(a) dx",
                "Rewrite: \\int \\frac{1}{u} \\cdot \\frac{1}{$(a)} du = \\frac{1}{$(a)} \\ln|u| + C",
                "Substitute back: \\frac{1}{$(a)}\\ln|$(tex(inner))| + C"
            ),
            time=120
        )
        
    else
        # Type 4: ∫ (ax+b)^n * cx dx where cx is NOT the derivative of (ax+b)
        # Requires algebraic manipulation before substitution
        a = choice([2, 3])
        b = randint(-8, 8)
        c = nonzero(-4, 4)
        n = choice([2, 3])
        
        inner = a*x + b
        integrand = c*x * inner^n
        
        # u = ax+b, x = (u-b)/a, dx = du/a
        # ∫ c*(u-b)/a * u^n * (1/a) du = (c/a²) ∫ (u^(n+1) - b*u^n) du
        # = (c/a²)[u^(n+2)/(n+2) - b*u^(n+1)/(n+1)] + C
        
        term1_coef = c // (a^2 * (n+2))
        term2_coef = -c*b // (a^2 * (n+1))
        antideriv = term1_coef * inner^(n+2) + term2_coef * inner^(n+1)
        
        problem(
            question="Evaluate \\int $(tex(integrand)) \\, dx",
            answer=antideriv,
            difficulty=(1000, 1200),
            solution=steps(
                sol("Given", integrand),
                "Let u = $(tex(inner)), then x = \\frac{u - $(b)}{$(a)} and dx = \\frac{1}{$(a)} du",
                "Substitute: \\int $(c) \\cdot \\frac{u-$(b)}{$(a)} \\cdot u^{$(n)} \\cdot \\frac{1}{$(a)} du",
                "Expand and integrate term by term",
                sol("Substitute back", antideriv)
            ),
            time=180
        )
    end
end